core/http/src/uri/uri.rs

```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
 impl_uri_from!(Origin);
 impl_uri_from!(Authority);
 impl_uri_from!(Absolute);
+
+mod uri_serde {
+    #![cfg(feature = "serde")]
+    
+    pub use super::Uri;
+    use std::fmt;
+    use std::convert::TryFrom;
+    // TODO pick import names
+    use _serde::{Deserialize, Deserializer, Serialize, Serializer, de::{self, Visitor}};
+
+    // The serialize implementation depends on the fmt::Display implementation. However, this
+    // should be okay, and is likely the best option, since allocating a string is unavoidable, and
+    // this method should allocate exactly one string
+    impl<'a> Serialize for Uri<'a> {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+                S: Serializer {
+            serializer.serialize_str(&format!("{}", self))
+        }
+    }
+
+    struct UriVisitor;
+
+    impl<'de> Visitor<'de> for UriVisitor {
+        type Value = Uri<'de>;
+        fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "a URI")
+        }
+
+        fn visit_borrowed_str<E>(self, v: &'de str) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v).map_err(de::Error::custom)
+        }
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v.to_string()).map_err(de::Error::custom)
+        }
+        fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v).map_err(de::Error::custom)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for Uri<'de> {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+                D: Deserializer<'de> {
+            deserializer.deserialize_str(UriVisitor)
+        }
+    }
+    
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        
+        use std::convert::TryFrom;
+        use crate::{ext::IntoOwned, uri::{self, Origin, Authority, Absolute, Host}};
+        #[derive(Debug)]
+        enum TestError {
+            UriError(Box<uri::Error<'static>>),
+            JsonError(serde_json::Error),
+        }
+        impl<'a> From<uri::Error<'a>> for TestError {
+            fn from(e: uri::Error<'a>) -> Self {
+                Self::UriError(Box::new(e.into_owned()))
+            }
+        }
+        impl From<serde_json::Error> for TestError {
+            fn from(e: serde_json::Error) -> Self {
+                Self::JsonError(e)
+            }
+        }
+
+        // Test Helper function
+        // 'a sould almost always be 'static, but it may be useful to not require that
+        fn test_helper<'a>(raw: &str, json: &'a str, uri_obj: Uri<'a>) -> Result<(), TestError> {
```

Comment:
Using `serde_test`, we should be able to get ride of the second parameter. Or even without `serde_test`, the second parameter is just `format!("\"{}\"", raw);`.

Nevertheless, this should either be `#[track_caller]` or a macro so that test failures point to the correct line.

Reply:
I will be making it `#[track_caller]`, thanks.